oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/core/expression/OAuth2Expressions.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core.expression;
+
+import org.springframework.security.core.Authentication;
+
+/**
+ * A contract for the OAuth2 SpEL expressions entry point
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public interface OAuth2Expressions {
```

Comment:
This interface is missing some operations. For example:

`hasScope(String scope)`
`hasAnyScope(String... scopes)`
`hasExactScope(String... scopes)`

I'm not necessarily suggesting all 3 of these operations but we do need an operation(s) that operate on `scope`.

What other operations do you think we might need based on the spec?

Reply:
We left them out of this commit on purpose, not sure if you remember that slack convo.

Happy to restate here for the record, though.

At the time, we were not settled on whether or not we would have `hasScope`, though we were certainly leaning towards it. I began adding things that seemed relevant from `spring-security-oauth` and, as I understood it, the decision was made to just have this commit prove out the bean concept itself and then add other expressions later.

I believe there is a ticket spring-security/spring-projects#5227 where this work is going currently.

If I misunderstood, though, let me know and we can pivot.

Reply:
Do you still agree that `ClaimsAccessor` (or a subclass) would be a good place for a `getScope`? If so, I'd say we might as well [get that done](https://github.com/spring-projects/spring-security/issues/5241) and then leverage that here on the resource server side.

Reply:
`AbstractOAuth2AccessTokenAuthenticationToken`  now contains `hasScope` and allows for configuring the attribute name via `setScopeAttributeName` as it may be different between providers. I would just delegate to this method when calling from the expressions `@Bean`.

Reply:
> Do you still agree that ClaimsAccessor (or a subclass) would be a good place for a getScope?

Scope is coupled to an Access Token so it makes sense to keep it in `AbstractOAuth2AccessTokenAuthenticationToken` and delegate calls to `AbstractOAuth2AccessTokenAuthenticationToken.hasScope`

Reply:
To answer your question from earlier about what other expressions we need, I think we should just keep in mind that these expressions are a finer-grained version of verifiers. Generally speaking, what questions we can ask in a verifier, we should be able to ask in an expression, too, e.g.

```java
public void verify(Jwt jwt) {
    assertNotNull(jwt.getAudience());
}

// ... and

@PreAuthorize("@oauth2.attribute('aud') != null")
```
The only other thing that I would stump for would be methods that are clearer than the standard attribute names. Like, I think that the following is more readable:

```java
@PreAuthorize("@oauth2.audience().contains('my-audience')")
```
than

```java
@PreAuthorize("@oauth2.attribute('aud').contains('my-audience')")
```

I wouldn't mind a `scopes` method:

```java
@PreAuthorize("@oauth2.scopes().contains('my-scope')")
```

Reply:
`AbstractOAuth2AccessTokenAuthenticationToken` I'm not sure is the right place for the logic. It isn't just the name of the attribute but also the type (Collection vs String).

An accessor seems to be a nice home for this:

```
get claim as list:
    if string then
        split by space and return
    else if collection then
        return
```
at least, but perhaps also:
```
get scope:
    if has claim "scope" then
        get claim as list
    else has claim "scp" then
        get claim as list
```
Accessors are where we have been housing other projection-type logic, say for timestamps.

Reply:
> I think that the following is more readable:  `@PreAuthorize("@oauth2.audience().contains('my-audience')")`

Totally agreed, but we cannot assume all access tokens will have an `aud` claim. Yes, JWT defines the `aud` claim as well as [Token Introspection](https://tools.ietf.org/search/rfc7662#section-2.2). 

How do you propose we handle cases where the Access Token does not have an `aud` claim?

Reply:
> `AbstractOAuth2AccessTokenAuthenticationToken` I'm not sure is the right place for the logic. It isn't just the name of the attribute but also the type (Collection vs String).

Maybe I didn't communicate my points clearly.

The current method in `AbstractOAuth2AccessTokenAuthenticationToken` is:

`public final boolean hasScope(String scope)`

It simply checks if the scope is available and returns true or false. I agree the internal method implementation should be changed to use the `ClaimAccessor` pattern as already discussed.

I also feel we can add another method to `AbstractOAuth2AccessTokenAuthenticationToken`:

`public Set<String> getScopes()` (same as in ``OAuth2AccessToken)

This internal method implementation would also leverage the `ClaimAccessor` pattern.

Does this explanation bring us on the same page?

Reply:
I think so. Let me try and apply what I've understood in an updated PR, and we can continue from there.

As for a missing audience, I think that is up to the caller. `audience()` would return null if there is none, in the same way that `attribute('aud')` would return null.